datafusion/physical-plan/src/recursive_query.rs

```diff
@@ -0,0 +1,363 @@
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Defines the recursive query plan
+
+use std::any::Any;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+
+use super::expressions::PhysicalSortExpr;
+use super::metrics::BaselineMetrics;
+use super::RecordBatchStream;
+use super::{
+    metrics::{ExecutionPlanMetricsSet, MetricsSet},
+    work_table::{WorkTable, WorkTableExec},
+    SendableRecordBatchStream, Statistics,
+};
+
+use arrow::datatypes::SchemaRef;
+use arrow::record_batch::RecordBatch;
+use datafusion_common::tree_node::{Transformed, TreeNode};
+use datafusion_common::{DataFusionError, Result};
+use datafusion_execution::TaskContext;
+use datafusion_physical_expr::Partitioning;
+use futures::{ready, Stream, StreamExt};
+
+use crate::{DisplayAs, DisplayFormatType, ExecutionPlan};
+
+/// Recursive query execution plan.
+///
+/// This plan has two components: a base part (the static term) and
+/// a dynamic part (the recursive term). The execution will start from
+/// the base, and as long as the previous iteration produced at least
+/// a single new row (taking care of the distinction) the recursive
+/// part will be continuously executed.
+///
+/// Before each execution of the dynamic part, the rows from the previous
+/// iteration will be available in a "working table" (not a real table,
+/// can be only accessed using a continuance operation).
+///
+/// Note that there won't be any limit or checks applied to detect
+/// an infinite recursion, so it is up to the planner to ensure that
+/// it won't happen.
+#[derive(Debug)]
+pub struct RecursiveQueryExec {
+    /// Name of the query handler
+    name: String,
+    /// The working table of cte
+    work_table: Arc<WorkTable>,
+    /// The base part (static term)
+    static_term: Arc<dyn ExecutionPlan>,
+    /// The dynamic part (recursive term)
+    recursive_term: Arc<dyn ExecutionPlan>,
+    /// Distinction
+    is_distinct: bool,
+    /// Execution metrics
+    metrics: ExecutionPlanMetricsSet,
+}
+
+impl RecursiveQueryExec {
+    /// Create a new RecursiveQueryExec
+    pub fn try_new(
+        name: String,
+        static_term: Arc<dyn ExecutionPlan>,
+        recursive_term: Arc<dyn ExecutionPlan>,
+        is_distinct: bool,
+    ) -> Result<Self> {
+        // Each recursive query needs its own work table
+        let work_table = Arc::new(WorkTable::new());
+        // Use the same work table for both the WorkTableExec and the recursive term
+        let recursive_term = assign_work_table(recursive_term, work_table.clone())?;
```

Comment:
Is it possible to have another recursive CTE nested within the recursive_term? If it is possible, we may need to ensure that we do not modify the work table of another CTE.

Reply:
Let's add a test to confirm

Reply:
nice catch - will add a test this afternoon.

Reply:
I ran out of time today :/ 

I plan to tackle this tomorrow.

Reply:
This is possible, and the result of the following SQL is different from PostgreSQL.
```sql
WITH RECURSIVE outer_cte AS (
    SELECT 1 as a
    UNION ALL (
        WITH  RECURSIVE nested_cte AS (
           SELECT 1 as a
           UNION ALL
           SELECT a+2 as a
	   FROM nested_cte where a < 3
         )
    SELECT outer_cte.a +2
    FROM outer_cte JOIN nested_cte USING(a)
    WHERE nested_cte.a < 4
   )
)
SELECT a FROM outer_cte;
```
Maybe we can choose not to support this case for now, and perform a check when connecting the work table.
```rust
fn assign_work_table(
    plan: Arc<dyn ExecutionPlan>,
    work_table: Arc<WorkTable>,
) -> Result<Arc<dyn ExecutionPlan>> {
    plan.transform_down(&|plan| {
        if let Some(exec) = plan.as_any().downcast_ref::<WorkTableExec>() {
            Ok(Transformed::Yes(Arc::new(
                exec.with_work_table(work_table.clone()),
            )))
        } else if plan.as_any().is::<RecursiveQueryExec>() {
            not_impl_err!("Recursive queries cannot be nested")
        } else {
            Ok(Transformed::No(plan))
        }
    })
}
```

Reply:
implemented here [`21e8a4e` (#8840)](https://github.com/apache/arrow-datafusion/pull/8840/commits/21e8a4e5d15469f28bde155cab78027d88bab308)